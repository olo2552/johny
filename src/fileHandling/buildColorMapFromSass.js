const buildColorMapFromSass = sassStreamFilePath => (
    sassStreamFilePath
        .trim()
        .split('\n')
        .filter(line => line.includes('#'))
        .map(line => line.split(':'))
        .map(line => line.map(line => line.trim()))
        .map(line => [line[0].slice(1), line[1].slice(0, -1)])
        .map(line => [line[0], line[1].toUpperCase()])
        .map(line => line.reverse())
        .reduce((colorsObj, line) => Object.assign(colorsObj, {
            [line[0]]: line[1],
        }), {})
);

module.exports = {
    buildColorMapFromSass
};
