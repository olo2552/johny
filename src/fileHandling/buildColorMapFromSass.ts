function buildColorMapFromSass(sassStream: string): object {
  return sassStream
    .trim()
    .split("\n")
    .filter(line => line.includes("#"))
    .map(line => line.split(":"))
    .map(line => line.map(lineElement => lineElement.trim()))
    .map(line => [line[0].slice(1), line[1].slice(0, -1)])
    .map(line => [line[0], line[1].toUpperCase()])
    .map(line => line.reverse())
    .reduce(
      (colorsObj, [colorName, color]) => ({
        ...colorsObj,
        [colorName]: color
      }),
      {}
    );
}

export { buildColorMapFromSass };
