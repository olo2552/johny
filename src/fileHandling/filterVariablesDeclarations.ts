import { IParseTree, nodeType } from "gonzales-pe";

function filterVariablesDeclarations(rootNode: IParseTree): IParseTree[] {
  const colors: IParseTree[] = [];

  rootNode.traverseByType("declaration", node => {
    colors.push(node);
  });

  return colors.filter(filteredDeclarationNode => {
    const variableType: nodeType = "variable";
    const property = filteredDeclarationNode.content[0];
    return property.content[0].type === variableType;
  });
}

export { filterVariablesDeclarations };
