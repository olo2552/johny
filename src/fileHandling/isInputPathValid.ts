import { access } from "fs";
import { extname } from "path";
import { promisify } from "util";
import { SUPPORTED_FILE_EXTENTIONS } from "../CONSTANTS";

async function isInputPathValid(path: string): Promise<boolean> {
  const fileExtension = extname(path);

  if (!SUPPORTED_FILE_EXTENTIONS.includes(fileExtension)) {
    throw Error(
      `Given file has unsupported format: "${fileExtension}"
Supported file extensions: ${SUPPORTED_FILE_EXTENTIONS.join(", ")}`
    );
  }

  try {
    await promisify(access)(path);
  } catch {
    throw Error(`Given file "${path}" does not exists.`);
  }

  return true;
}

export { isInputPathValid };
