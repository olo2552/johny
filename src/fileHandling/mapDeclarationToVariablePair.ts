import { IParseTree } from "gonzales-pe";
import { POSSIBLE_COLOR_TYPES } from "../constants";
import { IVariableWithColorNode } from "../interfaces";

function mapDeclarationToVariablePair(
  declarationNode: IParseTree
): IVariableWithColorNode {
  declarationNode.traverseByTypes(
    ["space", "propertyDelimiter"],
    (_, index, parent) => {
      parent.removeChild(index);
    }
  );

  const variableWithColor: IVariableWithColorNode = {
    color: declarationNode.content[1],
    variableName: ""
  };

  declarationNode.content[1].traverseByTypes(POSSIBLE_COLOR_TYPES, node => {
    variableWithColor.color = node;
  });

  declarationNode.content[0].traverseByType("ident", node => {
    variableWithColor.variableName = node.content;
  });

  return variableWithColor;
}

export { mapDeclarationToVariablePair };
