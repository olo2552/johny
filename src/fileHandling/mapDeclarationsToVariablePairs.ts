import { IParseTree } from "gonzales-pe";
import { IVariableWithColorNode } from "../interfaces";

function mapDeclarationsToVariablePairs(
  declarations: IParseTree[]
): IVariableWithColorNode[] {
  declarations.forEach(declaration => {
    declaration.traverseByTypes(
      ["space", "propertyDelimiter"],
      (_, index, parent) => {
        parent.removeChild(index);
      }
    );
  });

  return declarations.map(declaration => ({
    color: declaration.content[1],
    variable: declaration.content[0].content[0].content[0].content
  }));
}

export { mapDeclarationsToVariablePairs };
