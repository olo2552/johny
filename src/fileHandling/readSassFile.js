const { promisify } = require('util');
const { readFile } = require('fs');

const readSassFile = sassStreamFilePath =>
    promisify(readFile)(sassStreamFilePath, {
        encoding: 'utf-8',
        flag: 'r',
    })
        .catch(console.log);

module.exports = {
    readSassFile,
};