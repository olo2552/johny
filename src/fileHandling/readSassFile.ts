import { readFile } from "fs";
import { promisify } from "util";

function readSassFile(sassStreamFilePath: string): Promise<string> {
  return promisify(readFile)(sassStreamFilePath, {
    encoding: "utf-8",
    flag: "r"
  }).catch(err => {
    throw err;
  });
}

export { readSassFile };
