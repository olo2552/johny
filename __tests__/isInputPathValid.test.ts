import { appendFile, unlink } from "fs";
import { promisify } from "util";
import { isInputPathValid } from "../src/fileHandling/isInputPathValid";

describe("isInputPathValid", () => {
  const testFilePath = "./test.scss";

  beforeAll(async () => {
    await promisify(appendFile)(testFilePath, "");
  });

  it("should work correctly", async () => {
    expect.assertions(1);
    await expect(isInputPathValid(testFilePath));

    expect(true).toBe(true);
  });

  it("should reject file with incorrect extension", async () => {
    expect.assertions(1);

    await expect(
      isInputPathValid("kappa.txt")
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should accept file with supported extension", async () => {
    expect.assertions(1);
    await expect(isInputPathValid(testFilePath)).resolves.toBe(true);
  });

  it("should reject path guiding to file that does not exist", async () => {
    // expect.assertions(1);

    await expect(
      isInputPathValid("./notExists.scss")
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  afterAll(async () => {
    await promisify(unlink)(testFilePath);
  });
});
