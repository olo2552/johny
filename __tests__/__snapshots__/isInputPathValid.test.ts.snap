// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isInputPathValid should reject file with incorrect extension 1`] = `
"Given file has unsupported format: \\".txt\\"
Supported file extensions: .css, .sass, .scss, .less"
`;

exports[`isInputPathValid should reject path guiding to file that does not exist 1`] = `"Given file \\"./notExists.scss\\" does not exists."`;
